File: src/iter-4(~2023-10-31)/121. Best Time to Buy and Sell Stock/p2.py
Input = [7, 1, 5, 3, 6, 4]


def maxProfit(prices):
    max_num = 0
    left, right = 0, 1
    while right < len(prices):
        if prices[left] >= prices[right]:
            left = right
        else:
            max_num = max(max_num, prices[right] - prices[left])

        right += 1
    return max_num

File: src/iter-13(~2024-09-30)/121. Best Time to Buy and Sell Stock/p2.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  // base case
  if (prices.length === 1) {
    return 0;
  }
  let bestPrice = 0;
  let minNum = Infinity;

  for (const price of prices) {
    console.log({ price });
    if (minNum > price) {
      minNum = price;
      continue;
    }
    bestPrice = Math.max(bestPrice, price - minNum);
  }

  return bestPrice;
};

maxProfit([7, 1, 5, 3, 6, 4]); // 5
